include/mvdtool/mvd_generic.hpp

```diff
@@ -64,16 +64,16 @@ inline std::shared_ptr<File> open(const std::string& filename,
     case MVDType::MVD2:
         throw std::runtime_error("MVD2 not supported in the new MVD API");
     case MVDType::MVD3:
+        mvdfile.reset(new MVD3::MVD3File(filename));
         try {
-            mvdfile.reset(new MVD3::MVD3File(filename));
             mvdfile->size();
         } catch (...) {
             throw std::runtime_error("invalid MVD3 file");
         }
         break;
     default:
+        mvdfile.reset(new SonataFile(filename, population));
```

Comment:
This works here, because it is the last element. But to get the "invalid SONATA file" exception, it should actually be within the `try`… the constructor here reads the file structure already.

Reply:
Indeed. And propagating the constructors own exceptions seems adequate

Reply:
Now you're propagating one exception, but catching the other one. That seems inconsistent.

Reply:
Or you mean there might be unhandled cases ? 🤔